### ISSUE TITLE
Touchscreen support meta-issue

### ISSUE THREAD

**Issue by ids1024** [pop-os/cosmic-epoch#241]
Cosmic-comp has support for `wl_touch`, passing through touchscreen events from libinput to clients that handle `wl_touch`.

That's the main thing that's needed to make a touchscreen on a laptop somewhat useful, but there are more things that are needed for full touchscreen support. Especially for use of tablets without keyboards.

* On-screen keyboard
* Touch support in iced-sctk
* Touch support in xdg-shell-wrapper/cosmic-panel
* Touch support for SSD and other UI in cosmic-comp
   - Started in https://github.com/pop-os/cosmic-comp/pull/386
   - Moving and resizing windows with touchscreen

First-class support for touchscreen-only devices isn't the highest priority, but it's something to consider.

**Comment by ids1024** [comment]
Touch input was added to iced-sctk in https://github.com/pop-os/iced/pull/135.

For cosmic-panel, this will require https://github.com/pop-os/xdg-shell-wrapper/pull/22 and https://github.com/pop-os/cosmic-panel/pull/199.

**Comment by ahoneybun** [comment]
From testing the cosmic-panel PR and I think that pulls in the xdg-shell-wrapper one the only issues so far are:

-  able to move COSMIC applications (I can move GTK and others either way). 
- Using the toggles in Applets (you can trigger certain ones like opening Settings from the User Applet).

**Comment by ninjaquick** [comment]
Touch inputs do not map to the correct display (I can create a new item if this is something that's expected to be working):

- Display: PLANAR PCT2785
- Touch Device (lsusb): ID: 2575:0402 Weida Hi-Tech Co., LTD CooTouchR System
[lsusb.txt](https://github.com/user-attachments/files/16567313/lsusb.txt)

Inputs are detected but map to coordinates on the primary display.

Behaviors to compare to:

- Pop!OS_ 22.04: Input is mapped to the correct display automatically.
- VanillaOS 2 (Debian): Input is mapped to the correct display automatically. 
- KDE: Inputs don't appear to map automatically, but can be assigned via the Settings app.

**Comment by Vixeliz** [comment]
I don't know if this would be out of the scope and stay seperate (ie like lisgd). But having remappable touchscreen gestures would also be very cool

**Comment by Quackdoc** [comment]
@ninjaquick missed this a while ago, you can bind to specific display.

Get device name: `libinput list-devices | grep Device\:`
...
Device:           ILITEK ILITEK-TP
...

```ron
.config/cosmic/com.system76.CosmicComp/v1/input_devices
----------------------------
{
 "ILITEK ILITEK-TP": (
    state: Enabled,
    map_to_output: Some("HDMI-A-2")
    ),
}
```

**Comment by ThomasCorbin** [comment]
> I don't know if this would be out of the scope and stay seperate (ie like lisgd). But having remappable touchscreen gestures would also be very cool

KDE allows you to configure single finger gestures from each edge of the screen, which enables my favorite workflow, like triggering the kde 6 alt-tab behavior on once edge, and kde overview on another.   it works *great*

**Comment by vpoguru** [comment]
> [@ninjaquick](https://github.com/ninjaquick) missed this a while ago, you can bind to specific display.
> 
> Get device name: `libinput list-devices | grep Device\:` ... Device: ILITEK ILITEK-TP ...
> 
> .config/cosmic/com.system76.CosmicComp/v1/input_devices
> ----------------------------
> {
>  "ILITEK ILITEK-TP": (
>     state: Enabled,
>     map_to_output: Some("HDMI-A-1")
>     ),
> }

I have a iiyama touchscreen. 

```xrandr --listmonitors```shows the following information:

```shell
Monitors: 1
 0: +HDMI-A-1 1920/480x1080/270+0+0  HDMI-A-1
```

```libinput list-devices | grep Device\:``` shows the following information:

```
Device:           Power Button
Device:           Power Button
Device:           Cherry USB keyboard
Device:           Cherry USB keyboard System Control
Device:           Cherry USB keyboard Consumer Control
Device:           PixArt USB Optical Mouse
Device:           Weida Hi-Tech                CoolTouchR System            Mouse
Device:           Weida Hi-Tech                CoolTouchR System
```
The file ```.config/cosmic/com.system76.CosmicComp/v1/input_devices``` was not present, so I created one. I attempted to add the following configuration:

```ron
{
 "Weida Hi-Tech                CoolTouchR System            Mouse": (
    state: Enabled,
    map_to_output: Some("HDMI-A-1")
    ),
}
```
However, this configuration was unsuccessful. I also tried ```Weida Hi-Tech                CoolTouchR System``` and ```Weida Hi-Tech```, but neither of these configurations were successful. What am I missing?

**Comment by kajusnau** [comment]
Any updates on touch input support in `cosmic-panel`?
Would be cool to have this in ready for Alpha 8

**Comment by leviport** [comment]
@kajusnau No updates at this time. Please do not ping individual team members for this sort of vague request.

**Comment by denki** [comment]
> Any updates on touch input support in `cosmic-panel`? Would be cool to have this in ready for Alpha 8

It seems to be working now on `master`. See this PR: https://github.com/pop-os/cosmic-panel/pull/422

**Comment by ids1024** [comment]
Yeah. There may still be some details to fix, but touch support in panel applets is now generally working after https://github.com/pop-os/cosmic-panel/pull/422 and https://github.com/pop-os/iced/pull/234.

**Comment by mariayuno** [comment]
OSK support ? How is it going ? Any development ?

**Comment by ahoneybun** [comment]
> OSK support ? How is it going ? Any development ?

There isn't much going on in this repo currently so maybe it will be more of an Epoch 2 task:

https://github.com/pop-os/cosmic-osk

**Comment by FedX-sudo** [comment]
Unfortunately, this will take a very long time to implement properly and for me is the biggest limiting factor of COSMIC adoption. Namely the fact that the cursor does not hide on touchscreen inputs is fairly annoying compared to other desktops and compositors. 

I would also like to add that some devices such as my Framework 12 will send a tablet mode signal, which KDE will recognize to re-scale status bars for better touch mode inputs. However, I find the KDE behavior meh at best, specifically because the OSK will not toggle when switching modes and the behavior seems unconfigurable. 

I would really love to get in on implementing this if I ever find the time!

**Comment by ju-gem** [comment]
I'm not sure this is the right place for this, but palm rejection when using a pen/stylus would be essential for me to use cosmic with a touchscreen/two-in-one laptop.

**Comment by savuori** [comment]
Also, it would be great to be able to just disable the touchscreen.

**Comment by KAIYOHUGO** [comment]
> I'm not sure this is the right place for this, but palm rejection when using a pen/stylus would be essential for me to use cosmic with a touchscreen/two-in-one laptop.
> 

I think palm rejection can be configured in libinput 

ref: https://wayland.freedesktop.org/libinput/doc/latest/palm-detection.html

Edit: sorry for spanning, GitHub mobile app bug out again

**Comment by Popaulol** [comment]
I am currently having the issue that my Touch screen is working "correctly" but on the wrong monitor, so my touch inputs on my external touch monitor trigger inputs on my laptop screen (that isn't touch).

Cosmic 1.7.0 on Arch Linux (all packages up to date)
Framework Laptop 16 (Ryzen Ai 300)
External Monitor: Verbatim PMT-14

When i was still using KDE i could change the monitor the touch screen is mapped to in the KDE Settings. (If i find time in the next couple of days i'll figure out how to do it from the commandline at least temporarily until there's a good solution on cosmic).

**Comment by Jingles52** [comment]
> I am currently having the issue that my Touch screen is working "correctly" but on the wrong monitor, so my touch inputs on my external touch monitor trigger inputs on my laptop screen (that isn't touch).

In `~/.config/cosmic/com.system76.CosmicComp/v1/input_devices`, do something like this:

```
{
 "Wacom Movink 13 Pen": (
    state: Enabled,
    map_to_output: Some("DP-2")
    ),
 "Wacom Movink 13 Finger": (
    state: Enabled,
    map_to_output: Some("DP-2")
    ),
}
```

**Comment by Popaulol** [comment]
> In `~/.config/cosmic/com.system76.CosmicComp/v1/input_devices`, do something like this:
@Jingles52 Doesn't work for me (even after restarting either both just cosmic or the entire device)
I've also tried Changing DP-2 to DP-1 since my external shows as screen 1 in the cosmic Settings.
Touch always went to the laptop internal screen.

 Do you also know how touch would interact with regularly switching between 3 different monitor setups (Even if i find a fix until then, I can only test my 3rd Setup in a couple of days since i'm not home rn)

On my home Setup i have 3 externals (one of which is touch) and my non touch internal display.
When I'm on the go, i either only have my laptop display or my laptop and the touch display.

Do i have to fuck with the config anytime i switch setups or does it "just work" once set up properly (like it did with KDE and Gnome when i was still using those)?

I also have an external wacom Tablet without a screen (ETA-0405A-U) that i use sometimes (both on the large setup and on the go, usually mapped to the touch screen), is mapping that done in a similar way to the touch screen, or are drawing tablets even supported at all rn?

**Comment by Jingles52** [comment]
Make sure you're using your names for inputs/outputs, ask your clanker how to enumerate them. From my experience, it justwerkz after you set up the config, no matter what subset of displays you have on.
